pkg/applications/helmclient/client.go

```diff
@@ -0,0 +1,311 @@
+	// release information will be stored in the targetNamespace which lead to a release we can not uninstall with helm.
+	if kcNamspace, _, err := restClientGetter.ToRawKubeConfigLoader().Namespace(); err != nil || kcNamspace != targetNamespace {
+		return nil, fmt.Errorf("namespace set in RESTClientGetter should be the same as targetNamespace. RESTClientGetter namespace=%s, targetNamespace=%s: %w", kcNamspace, targetNamespace, err)
+	}
+
+	actionConfig := new(action.Configuration)
+	if err := actionConfig.Init(restClientGetter, targetNamespace, secretStorageDriver, logger.Infof); err != nil {
+		return nil, fmt.Errorf("can not initialize helm actionConfig: %w", err)
+	}
+
+	return &HelmClient{
+		resClientGetter: restClientGetter,
+		settings:        settings,
+		getterProviders: settings.GetterProviders(),
+		actionConfig:    actionConfig,
+		targetNamespace: targetNamespace,
+		logger:          logger,
+	}, nil
+}
+
+// DownloadChart from url into dest folder and return the chart location (eg /tmp/foo/apache-1.0.0.tgz)
+// The dest folder must exist.
+func (h HelmClient) DownloadChart(url string, chartName string, version string, dest string) (string, error) {
+	var repoName string
+	var err error
+
+	if strings.HasPrefix(url, "oci://") {
+		repoName = url
+	} else {
+		repoName, err = h.ensureRepository(url)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	var out strings.Builder
+	chartDownloader := downloader.ChartDownloader{
+		Out:              &out,
+		Verify:           downloader.VerifyNever,
+		RepositoryConfig: h.settings.RepositoryConfig,
+		RepositoryCache:  h.settings.RepositoryCache,
+		Getters:          h.getterProviders,
+		// todo credentials
+	}
+	// todo note: we may want to check the verificaton return by chartDownloader.DownloadTo. for the moment its set to downloader.VerifyNever in struct init
+	chartRef := repoName + "/" + chartName
+	chartLoc, _, err := chartDownloader.DownloadTo(chartRef, version, dest)
+	if err != nil {
+		h.logger.Errorf("failed to download chart '%s' in version '%s'. log='%s'", chartRef, version, out.String())
+		return "", err
+	}
+
+	h.logger.Debugf("successful download of chart '%s' in version '%s'. log='%s'", chartRef, version, out.String())
+	return chartLoc, nil
+}
+
+// InstallOrUpgrade installs the chart located at chartLoc into targetNamespace if it's not already installed.
+// Otherwise it updates the chart.
+// charLoc is the path to the chart archive (eg /tmp/foo/apache-1.0.0.tgz) or folder containning the chart (eg /tmp/mychart/apache).
+func (h HelmClient) InstallOrUpgrade(chartLoc string, releaseName string, values map[string]interface{}) (*release.Release, error) {
+	if _, err := h.actionConfig.Releases.Last(releaseName); err != nil {
+		return h.Install(chartLoc, releaseName, values)
+	}
+	return h.Upgrade(chartLoc, releaseName, values)
+}
+
+// Install the chart located at chartLoc into targetNamespace if the chart is already installed, an error is returned.
```

Comment:
Please rephrase this a bit, it reads as "installs the chart if it's already installed" and "an error is returned", which is confusing :D A semicolon could help, like

> Install the chart located at chartLoc; if the chart is installed, an error is returned